File: src/data/types/assignResult.ts
import type { CrewMember } from "./crewMember.ts";
import type { CraftResult } from "./craftResult.ts";

// Necessary completion: Agent 16's contract names `assignToCraft(...):
// AssignResult` but never defines it.
export interface AssignSucceeded {
  assigned: true;
  updatedCrewMember: CrewMember;
  craftResult: CraftResult;
}

export interface AssignRejected {
  assigned: false;
  reason: string;
}

export type AssignResult = AssignSucceeded | AssignRejected;

File: unity/ProfitableCore/Schema/AssignResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/assignResult.ts's AssignResult = AssignSucceeded |
// AssignRejected discriminated union.
public abstract class AssignResult
{
    public abstract bool Assigned { get; }
}

public sealed class AssignSucceeded : AssignResult
{
    public override bool Assigned => true;
    public CrewMember UpdatedCrewMember { get; init; } = new();
    public CraftResult CraftResult { get; init; } = null!;
}

public sealed class AssignRejected : AssignResult
{
    public override bool Assigned => false;
    public string Reason { get; init; } = string.Empty;
}
